openquake/commonlib/readinput.py

```diff
@@ -912,14 +932,72 @@ def get_gmfs(oqparam, sitecol):
                     raise InvalidFile(
                         'The column #%d in %s is expected to contain positive '
                         'floats, got %s instead' % (i + 3, fname, row[i + 2]))
-                gmf_by_imt[imts[i]][lineno - 1, :] = r_sites.expand(array, 0)
+                gmf_by_imt[imts[i]][lineno - 1] = r_sites.expand(array, 0)
             tags.append(row[0])
     if lineno < num_gmfs:
         raise InvalidFile('%s contains %d rows, expected %d' % (
             fname, lineno, num_gmfs))
     if tags != sorted(tags):
         raise InvalidFile('The tags in %s are not ordered: %s' % (fname, tags))
-    return gmf_by_imt.T
+    return sitecol, numpy.array(tags, '|S100'), gmf_by_imt.T
+
+
+# used in get_scenario_from_nrml
+def _extract_tags_sites(gmfset):
+    tags = set()
+    mesh = set()
+    for gmf in gmfset:
+        tags.add(gmf['ruptureId'])
+        for node in gmf:
+            mesh.add((node['lon'], node['lat']))
+    return numpy.array(sorted(tags), '|S100'), sorted(mesh)
+
+
+def get_scenario_from_nrml(oqparam, fname):
+    """
+    :param oqparam:
+        an :class:`openquake.commonlib.oqvalidation.OqParam` instance
+    :param fname:
+        the NRML files containing the GMFs
+    :returns:
+        a triple (sitecol, rupture_tags, gmf array)
+    """
+    imts = list(oqparam.imtls)
+    imt_dt = numpy.dtype([(imt, float) for imt in imts])
+    gmfset = nrml.read(fname).gmfCollection.gmfSet
+    tags, oqparam.sites = _extract_tags_sites(gmfset)
+    oqparam.number_of_ground_motion_fields = e = len(tags)
+    sitecol = get_site_collection(oqparam)
+    n = len(oqparam.sites)
```

Comment:
The variables `e`, `n`, `m`, `i`, `j` in this function could potentially be renamed to have more reader friendly names.